### PR TITLE
MINOR: revisit LogValidatorTest#checkRecompression

### PR DESCRIPTION
**Pull request by chia7712**
from https://github.com/apache/kafka/pull/15904#discussion_r1598902252

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
ping @junrao @vincent81jiang

File: core/src/test/scala/unit/kafka/log/LogValidatorTest.scala [hunk 1]
```diff
@@ -497,9 +497,11 @@ class LogValidatorTest {
     assertEquals(now + 1, validatingResults.maxTimestampMs,
       s"Max timestamp should be ${now + 1}")
 
-    // Both V2 and V1 has single branch in the record when compression is enable, and hence their shallow OffsetOfMaxTimestamp
-    // is the last offset of the single branch
-    assertEquals(1, records.batches().asScala.size)
+    // V2 has single batch, and other versions has many single-record batches
+    assertEquals(if (magic >= RecordBatch.MAGIC_VALUE_V2) 1 else 3, records.batches().asScala.size)
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: Should we move this assert to immediate after records is created?

File: core/src/test/scala/unit/kafka/log/LogValidatorTest.scala [hunk 2]
```diff
@@ -497,9 +497,11 @@ class LogValidatorTest {
...
+    // Both V2 and V1 has single batch in the record when compression is enable, and hence their shallow OffsetOfMaxTimestamp
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: in the record => in the validated records

**Review by junrao — COMMENTED** [review]
@chia7712 : Thanks for the PR. Just a couple of minor comments.

**Comment by chia7712** [comment]
@junrao thanks for all your reviews :)

**Review by junrao — APPROVED** [review]
@chia7712 : Thanks for the updated PR. LGTM. 

The test failures seem unrelated, but it would be useful to file jiras for the untracked ones.

**Comment by chia7712** [comment]
|test|jira|
|-----|-----|
|testDelayedConfigurationOperations|https://issues.apache.org/jira/browse/KAFKA-16773|
|testOffsetTranslationBehindReplicationFlow|https://issues.apache.org/jira/browse/KAFKA-15197|
|testAlterSinkConnectorOffsets|https://issues.apache.org/jira/browse/KAFKA-15892|
|testElectPreferredLeaders|https://issues.apache.org/jira/browse/KAFKA-8458|
|testSeparateOffsetsTopic|https://issues.apache.org/jira/browse/KAFKA-14089|
|testNoConsumeWithoutDescribeAclViaSubscribe|https://issues.apache.org/jira/browse/KAFKA-9655|
|testNoConsumeWithDescribeAclViaSubscribe|https://issues.apache.org/jira/browse/KAFKA-7646|
|testCreateUserWithDelegationToken|https://issues.apache.org/jira/browse/KAFKA-16501|
|testDescribeTokenForOtherUserFails|https://issues.apache.org/jira/browse/KAFKA-15411|
|testCoordinatorFailover|https://issues.apache.org/jira/browse/KAFKA-16024|
|shouldWorkWithUncleanShutdownWipeOutStateStore|https://issues.apache.org/jira/browse/KAFKA-16502|
|shouldCloseAllTaskProducersOnCloseIfEosEnabled|https://issues.apache.org/jira/browse/KAFKA-16774|
|testTaskRequestWithOldStartMsGetsUpdated|https://issues.apache.org/jira/browse/KAFKA-8115|
|testSyncTopicConfigs|https://issues.apache.org/jira/browse/KAFKA-14453|
|testReplicateSourceDefault|https://issues.apache.org/jira/browse/KAFKA-15292|
|testCreateTokenForOtherUserFails|https://issues.apache.org/jira/browse/KAFKA-15411|
|testDescribeTokenForOtherUserFails|https://issues.apache.org/jira/browse/KAFKA-15411|
|testCreateExistingTopicsThrowTopicExistsException|https://issues.apache.org/jira/browse/KAFKA-16775|
|testLowMaxFetchSizeForRequestAndPartition|https://issues.apache.org/jira/browse/KAFKA-16494|
|testTaskRequestWithOldStartMsGetsUpdated|https://issues.apache.org/jira/browse/KAFKA-8115|

**Pull request merged**